Jellyfin.Server.Implementations/MediaSegments/MediaSegmentsManager.cs

```diff
@@ -0,0 +1,215 @@
+            using (dbContext)
+            {
+                IQueryable<MediaSegment> queryable = dbContext.Segments.Select(s => s);
+
+                if (!itemId.Equals(default))
+                {
+                    queryable = queryable.Where(s => s.ItemId.Equals(itemId));
+                }
+
+                if (!streamIndex.Equals(-1))
+                {
+                    queryable = queryable.Where(s => s.StreamIndex.Equals(streamIndex));
+                }
+
+                if (!type.Equals(null))
+                {
+                    queryable = queryable.Where(s => s.Type.Equals(type));
+                }
+
+                if (typeIndex > -1)
+                {
+                    queryable = queryable.Where(s => s.TypeIndex.Equals(typeIndex));
+                }
+
+                allSegments = queryable.AsNoTracking().ToList();
+            }
+
+            return allSegments;
+        }
+
+        /// <summary>
+        /// Add or Update a segment in db.
+        /// <param name="dbContext">The db context.</param>
+        /// <param name="segment">The segment.</param>
+        /// <param name="found">The found segment.</param>
+        /// </summary>
+        private void AddOrUpdateSegment(JellyfinDbContext dbContext, MediaSegment segment, MediaSegment? found)
+        {
+            if (found != null)
+            {
+                found.StartTicks = segment.StartTicks;
+                found.EndTicks = segment.EndTicks;
+                found.Action = segment.Action;
+            }
+            else
+            {
+                dbContext.Segments.Add(segment);
+            }
+        }
+
+        /// <summary>
+        /// Validate a segment: itemId, start >= end and type.
+        /// </summary>
+        /// <param name="segment">The segment to validate.</param>
+        private void ValidateSegment(MediaSegment segment)
+        {
+            if (segment.ItemId.Equals(default))
+            {
+                throw new ArgumentException($"itemId is default: itemId={segment.ItemId} for segment with type '{segment.Type}.{segment.TypeIndex}'");
+            }
+
+            if (segment.StartTicks >= segment.EndTicks)
+            {
+                throw new ArgumentException($"start >= end: {segment.StartTicks}>={segment.EndTicks} for segment itemId '{segment.ItemId}' with type '{segment.Type}.{segment.TypeIndex}'");
+            }
+        }
+
+        /// <summary>
+        /// Delete all segments when itemid is deleted from library.
+        /// TODO: Do not block.
+        /// </summary>
+        /// <param name="sender">The sending entity.</param>
+        /// <param name="itemChangeEventArgs">The <see cref="ItemChangeEventArgs"/>.</param>
+        private void LibraryManagerItemRemoved(object? sender, ItemChangeEventArgs itemChangeEventArgs)
+        {
+            var task = Task.Run(async () => { await DeleteSegmentsAsync(itemChangeEventArgs.Item.Id).ConfigureAwait(false); });
+            task.Wait();
```

Comment:
Why not async/await?

Reply:
I tried to use the following but the `EventHandler` is sync. Any idea how i can write that cleaner? I'm not into deep with c#
```
        private async Task LibraryManagerItemRemoved(object? sender, ItemChangeEventArgs itemChangeEventArgs)
        {
            await DeleteSegmentsAsync(itemChangeEventArgs.Item.Id).ConfigureAwait(false);
        }
```

Thank you

Reply:
Well the whole reason why `async void` exists is because of eventhandlers like those. I hate this implementation and would rather see a PubSub delegate handling but this is clearly out of scope here.

for the time being, `private async void LibraryManagerItemRemoved` will suffice.
A small note here tho, this changes the call structure a bit as the invoker of the event will continue its execution once you actually hit a blocking await call. I dont think this is an issue here tho, it just means that whoever raises the event, will continue with his code while `DeleteSegmentsAsync` might not be completed yet.

Reply:
Thank you!